src/finch/tensor.py

```diff
@@ -156,12 +159,20 @@ def __mod__(self, other):
     def __pow__(self, other):
         return self._elemwise_op("^", other)
 
-    def __matmul__(self, other: "Tensor") -> "Tensor":
-        if self.ndim < 2 or other.ndim < 2:
-            raise NotImplementedError(
-                "`@` is not implemented for arrays with less than two dimensions."
+    @compiled()
+    def __matmul__(self, other: Tensor) -> Tensor:
+        if self.ndim == 0 or other.ndim == 0:
+            raise ValueError(
+                f"`{self.ndim=}`, `{other.ndim=}`. Both must be greater than `0`."
             )
-        return tensordot(self, other, axes=((-1,), (-2,)))
+
+        if other.ndim == 1:
+            return sum(self * other, axis=-1)
+
+        if self.ndim == 1:
+            return sum(self * other.mT, axis=-1)
+
+        return sum(self[..., :, None, :] * other.mT[..., None, :, :], axis=-1)
```

Comment:
Is this all that was needed? That's quite clever!

Reply:
Yes! The lines above generalise this function to a vector-vector, matrix-vector and vector-matrix product, and this one is for a matrix-matrix product.